### DISCUSSION TITLE
Recompiling PyScipOPT after extending the interface

### DISCUSSION THREAD

**Discussion by jgroenheide** [scipopt/PySCIPOpt#787] [Q&A] [answered]
Hi,
I'm working on a project where I've added PySCIPOpt as a package through pip, but I needed to extend the interface with some additional functionality. I've added `getPrioChild()` and `getPrioSibling()` to the scip.pxi and scip.pxd files, but I'm struggling to get my interpreter to recognise the changes. 
```
    self.model.getPrioChild()
AttributeError: 'pyscipopt.scip.Model' object has no attribute 'getPrioChild'
  File "src\pyscipopt\scip.pxi", line 3355, in pyscipopt.scip.Model.optimize  <-- This is not the correct line for Model.optimize, showing that it is reading the wrong file
  File "src\pyscipopt\scip.pxi", line 256, in pyscipopt.scip.PY_SCIP_CALL
Exception: SCIP: unspecified error!
```
I'm using an unmodified version of SCIP, so I have not compiled my SCIP from source.
Any help would be greatly appreciated, and thank you in advance.

**Comment by Joao-Dionisio** [comment]
Hello, @jgroenheide! You need to go to the folder with the setup.py file and write ```pip install .```


Cython creates a C file when translating the python code, that's why you don't recognize the line :)

Also, please consider creating a PR with your changes, if you think it might be helpful!

**Reply by jgroenheide** [reply]
I imported PySCIPOpt using Pip, so I don't have the setup.py file in my project environment. Is there a way to add it, or do I need to install PySCIPOpt through a different method?

**Reply by Joao-Dionisio** [reply]
Ah, you're right @jgroenheide, I assumed that it would work as well. Hmm, I'm nearly positive that it wouldn't work, but you can try to copy the ```setup.py``` file that's here in the repo into the ```pyscipopt``` folder that you have. You can find its location with ```pip show pyscipopt```.

Alternatively, you can try to install it from source. I'm a newbie at this stuff, but here are the notes I took at the time

1. git clone https://github.com/scipopt/PySCIPOpt
2. mv PySCIPOpt [the name you want] (<- optional, I just changed the name to differentiate)
3. cd [the name you want]
4. Do your changes
5. In this folder, where ```setup.py``` is, do ```pip install .```

**Reply by jgroenheide** [reply]
I did try copying the setup.py file already, which didn't work. I have not found another way to recompile the package after it's been installed into my virtual environment, which means the only remaining option is probably to compile from source. Likewise, I'm also a newbie at that kind of stuff, however, so I was hoping to avoid it. Thank you for the feedback though. Maybe it's worth it to put these steps in the INSTALL.md file as well, once we find something that works.

**Reply by Joao-Dionisio** [reply]
@jgroenheide yeah, I've been delaying doing some tutorial videos for pyscipopt, some of them related to installation stuff.

Let me know if my steps work out! Forgot to mention, but the steps are to be taken in the terminal (also, they're for Ubuntu :/ )

**Comment by jgroenheide** [comment, marked as answer]
From what I can tell, the only way to extend the interface is by compiling it from source.

I performed the following steps on Windows.
1. Clone the repository to any local directory.
2. Edit the System environment variables by adding `SCIPOPTDIR=/path/to/scip/install/dir` as explained in [INSTALL.md](https://github.com/scipopt/PySCIPOpt/blob/master/INSTALL.md)
(I used Windows' built-in System Properties window for this.)
3. Open the virtual environment or project (with the virtual environment enabled) where you want to install the extended version of PySCIPOpt.
4. Navigate to the cloned repository using `cd` and install the package with `python -m pip install .`
6. Now you can make changes to the `scip.pxd` and `scip.pxi` files. I used VS Code for this without issues.
7. Recompiling the code can now be done using `python -m pip install .` or `python -m pip install --compile .` 
(I didn't notice a difference, but maybe someone more experienced can explain which one should be used.)

I did not need to install the Python development files, or use `export` for the environment variable. Looking back on it, the [INSTALL.md](https://github.com/scipopt/PySCIPOpt/blob/master/INSTALL.md) instructions for compiling from source do give all the correct commands. The instructions for how to extend the interface should probably mention that this only works when compiling PySCIPOpt from source, though this might be obvious to more experienced developers.

**Reply by Joao-Dionisio** [reply]
Very happy to hear this! As I've said, please do consider making a PR with your changes, if you believe it might benefit the community. Happy optimizing! :)
